File: maths/sieve_of_eratosthenes.go
package main

import (
	"fmt"
	"math"
)

func PrimeSieve(num int) []int {
	if num <= 0 {
		return []int{}
	}

	sieve := make([]bool, num+1)
	for i := range sieve {
		sieve[i] = true
	}
	var primes []int

	limit := int(math.Sqrt(float64(num)))

	for p := 2; p <= limit; p++ {
		if sieve[p] {
			for i := p * p; i <= num; i += p {
				sieve[i] = false
			}
		}
	}

	for p := 2; p <= num; p++ {
		if sieve[p] {
			primes = append(primes, p)
		}
	}
	return primes
}

func main() {
	n := 50
	fmt.Printf("Primes up to %d: %v\n", n, PrimeSieve(n))
}

File: maths/sieve_of_eratosthenes.cpp
#include <cmath>
#include <iostream>
#include <vector>

std::vector<int> primeSieve(int num) {
  if (num <= 0)
    return {};

  std::vector<bool> sieve(num + 1, true);
  std::vector<int> primes;
  int start = 2;
  int end = std::sqrt(num);

  for (int p = 2; p * p <= num; p++) {
    // If sieve[p] is not changed, then it is a prime
    if (sieve[p] == true) {
      // Update all multiples of p
      for (int i = p * p; i <= num; i += p)
        sieve[i] = false;
    }
  }

  // Print all prime numbers
  for (int p = 2; p <= num; p++)
    if (sieve[p])
      primes.push_back(p);

  return primes;
}

int main() {
  int n = 50;
  std::vector<int> primes = primeSieve(n);
  std::cout << "Primes up to " << n << ": ";
  for (int p : primes)
    std::cout << p << " ";
  std::cout << std::endl;
  return 0;
}
